File: benchmark/synduce/tree/sumtree.f
Inductive Tree = nil Unit | node {Int, Tree, Tree};

spec = (fix (
  \f: Int -> Tree -> Int. \w: Int. \t: Tree.
  match t with
    nil _ -> w
  | node {a, l, r} -> f (+ (f w l) a) r
  end
)) 0;

repr = fix (
  \f: Tree -> Compress Tree. \t: Tree.
  match t with
    nil _ -> nil unit
  | node {a, l, r} -> node {a, f l, f r}
  end
);

main = \t: Tree. spec (repr t);

File: ml-benchmark/synduce/tree/sumtree.ml
type tree = Nil | Node of int * tree * tree

let rec f w t =
  match t with
  | Nil -> w
  | Node (a, l, r) ->
      let wl = f w l in
      let w2 = wl + a in
      f w2 r

let spec t =
  f 0 t

let rec repr t =
  match t with
  | Nil -> Nil
  | Node (a, l, r) ->
      let l2 = repr l in
      let r2 = repr r in
      Node (a, l2, r2)

let program t =
  spec (repr t)